### PR TITLE
Auto-generated Schemas documentation 

### PR DESCRIPTION
**Pull request by ammirate**
This PR integrate the project jsonschema2rst inside inspire-schemas, allowing to auto generate schemas-documentation on the fly. The generated docs in linked by the main documentation page.

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -155,10 +180,7 @@
 # Theme options are theme-specific and customize the look and feel of a theme
 # further.  For a list of options available for each theme, see the
 # documentation.
-#html_theme_options = {}
-
-# Add any paths that contain custom themes here, relative to this directory.
-#html_theme_path = []
```
Comment (david-caro) [inline docs/conf.py]: Not that it's an issue or anything, but there's no reason to move this right? Probably a failed try?

**Comment by david-caro** [comment]
You also missed adding `(closes #73)` at the end of the commit message somewhere.

File: docs/conf.py (hunk 1 above)
Comment (ammirate) [inline docs/conf.py]: You are right!

File: setup.py [hunk 2]
```diff
@@ -94,7 +93,7 @@ def do_setup(url=URL):
         zip_safe=False,
         autosemver=True,
     )
-
+    _generate_json_schemas()
```
Comment (michamos) [inline setup.py]: @ammirate maybe this change makes the build flaky?

**Comment by michamos** [comment]
@ammirate I think you rebased master on top of this, instead of the other way around.

File: setup.py (hunk 2 above)
Comment (michamos) [inline setup.py]: It is actually required, because otherwise `import yaml` happens before installing `pyyaml`.

**Comment by michamos** [comment]
@david-caro it should be ready now

File: setup.py [hunk 3]
```diff
@@ -94,6 +93,7 @@ def do_setup(url=URL):
         zip_safe=False,
         autosemver=True,
     )
+    _generate_json_schemas()
```
Comment (david-caro) [inline setup.py]: shouldn't this go before setup? so the json files are there when the setup packages?

File: setup.py [hunk 4]
```diff
@@ -71,8 +71,7 @@ def _generate_json_schemas():
         _yaml2json(yaml_file=yaml_file, json_file=json_file)
 
 
-def do_setup(url=URL):
-    _generate_json_schemas()
+def do_setup():
```
Comment (david-caro) [inline setup.py]: I think that the issue with the url param is that it was not replaced down in the setup function, it's a "bad habit" to use global variables. Not a big deal here though.

File: setup.py (hunk 3 above)
Comment (david-caro) [inline setup.py]: Just tested, if you don't install the package (that is, run `pip install -e .` or run any other setup command, then the json files are not added to the dist):

```shell
01:51 PM ~/Work/repos/per_user/inspirehep/inspire-schemas<venv:inspire-schemas>  (master ↑·2|✔) 
dcaro@pcrcssis001$ python setup.py sdist
/home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages/autosemver/__init__.py:156: UserWarning: Using explicit parametes is deprecated and will be removed, pass a dictionary as the value of the autosemver parameter instead.
  'Using explicit parametes is deprecated and will be removed, '
running sdist
running egg_info
creating inspire_schemas.egg-info
writing requirements to inspire_schemas.egg-info/requires.txt
writing inspire_schemas.egg-info/PKG-INFO
writing top-level names to inspire_schemas.egg-info/top_level.txt
writing dependency_links to inspire_schemas.egg-info/dependency_links.txt
writing manifest file 'inspire_schemas.egg-info/SOURCES.txt'
reading manifest file 'inspire_schemas.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'inspire_schemas.egg-info/SOURCES.txt'
running check
creating inspire-schemas-40.0.2
creating inspire-schemas-40.0.2/inspire_schemas
creating inspire-schemas-40.0.2/inspire_schemas.egg-info
creating inspire-schemas-40.0.2/inspire_schemas/records
creating inspire-schemas-40.0.2/inspire_schemas/records/elements
copying files to inspire-schemas-40.0.2...
copying AUTHORS -> inspire-schemas-40.0.2
copying CHANGELOG -> inspire-schemas-40.0.2
copying MANIFEST.in -> inspire-schemas-40.0.2
copying README.rst -> inspire-schemas-40.0.2
copying setup.cfg -> inspire-schemas-40.0.2
copying setup.py -> inspire-schemas-40.0.2
copying inspire_schemas/__init__.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas/api.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas/builders.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas/errors.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas/utils.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas.egg-info/PKG-INFO -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/SOURCES.txt -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/dependency_links.txt -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/not-zip-safe -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/requires.txt -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/top_level.txt -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas/records/__init__.py -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/elements/__init__.py -> inspire-schemas-40.0.2/inspire_schemas/records/elements
Writing inspire-schemas-40.0.2/setup.cfg
creating dist
Creating tar archive
removing 'inspire-schemas-40.0.2' (and everything under it)

01:51 PM ~/Work/repos/per_user/inspirehep/inspire-schemas<venv:inspire-schemas>  (master ↑·2|✔) 
dcaro@pcrcssis001$ python setup.py sdist
/home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages/autosemver/__init__.py:156: UserWarning: Using explicit parametes is deprecated and will be removed, pass a dictionary as the value of the autosemver parameter instead.
  'Using explicit parametes is deprecated and will be removed, '
running sdist
running egg_info
writing requirements to inspire_schemas.egg-info/requires.txt
writing inspire_schemas.egg-info/PKG-INFO
writing top-level names to inspire_schemas.egg-info/top_level.txt
writing dependency_links to inspire_schemas.egg-info/dependency_links.txt
reading manifest file 'inspire_schemas.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'inspire_schemas.egg-info/SOURCES.txt'
running check
creating inspire-schemas-40.0.2
creating inspire-schemas-40.0.2/inspire_schemas
creating inspire-schemas-40.0.2/inspire_schemas.egg-info
creating inspire-schemas-40.0.2/inspire_schemas/records
creating inspire-schemas-40.0.2/inspire_schemas/records/elements
copying files to inspire-schemas-40.0.2...
copying AUTHORS -> inspire-schemas-40.0.2
copying CHANGELOG -> inspire-schemas-40.0.2
copying MANIFEST.in -> inspire-schemas-40.0.2
copying README.rst -> inspire-schemas-40.0.2
copying setup.cfg -> inspire-schemas-40.0.2
copying setup.py -> inspire-schemas-40.0.2
copying inspire_schemas/__init__.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas/api.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas/builders.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas/errors.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas/utils.py -> inspire-schemas-40.0.2/inspire_schemas
copying inspire_schemas.egg-info/PKG-INFO -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/SOURCES.txt -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/dependency_links.txt -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/not-zip-safe -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/requires.txt -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas.egg-info/top_level.txt -> inspire-schemas-40.0.2/inspire_schemas.egg-info
copying inspire_schemas/records/__init__.py -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/authors.json -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/conferences.json -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/data.json -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/experiments.json -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/hep.json -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/institutions.json -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/jobs.json -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/journals.json -> inspire-schemas-40.0.2/inspire_schemas/records
copying inspire_schemas/records/elements/__init__.py -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/acquisition_source.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/address.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/arxiv_categories.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/cnum.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/contact.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/country_code.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/degree_type.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/document_type.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/id.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/inspire_field.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/json_reference.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/language_code.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/material.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/rank.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/reference.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/related_record.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/source.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/sourced_value.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/title.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
copying inspire_schemas/records/elements/url.json -> inspire-schemas-40.0.2/inspire_schemas/records/elements
Writing inspire-schemas-40.0.2/setup.cfg
Creating tar archive
removing 'inspire-schemas-40.0.2' (and everything under it)
```

Comment (michamos) [inline setup.py]: It cannot go before, as it tries to `import yaml` but that dependency is not necessarily installed yet.
The problem you raised is what makes Travis fail on Python 3, even though it should `pip install -e .` first, but works fine on Python 2.

**Comment by michamos** [comment]
@david-caro I put the Python 3 build back and reordered the `do_setup`

**Comment by david-caro** [comment]
Wait, this generates a duplicated documentation, one for the yml and another for the json.

**Comment by michamos** [comment]
Not here: http://inspire-schemas.readthedocs.io/en/docstests/schemas/index.html

**Comment by david-caro** [comment]
That one is not this patch, but master.

**Comment by michamos** [comment]
No, it's this branch, `docstests`

**Comment by david-caro** [comment]
True, not sure why it's not duplicated there then, if you run locally you get it twice.

**Comment by david-caro** [comment]
I think that the difference is that it runs the spinx command from the ./docs directory, instead of the root of the repo.

**Comment by michamos** [comment]
Locally I also see the schema docs only once. `jsonschema2rst` runs twice though, once for `html` and once for `doctrees`.

**Comment by michamos** [comment]
@david-caro so, OK to merge?

**Comment by david-caro** [comment]
Let's try to figure out why I see it twice first...

running in the docs dir instead of the root does not work for me, and running `python setup.py install` instead of `pip install -e .` has no change either.

What are you running to generate the docs?

**Comment by david-caro** [comment]
This is what I get:
```shell
09:37 AM ~/Work/repos/per_user/inspirehep/inspire-schemas<venv:inspire-schemas>  (master ↑·3|✔) 
dcaro@pcrcssis001$ git log -1
commit 507984eacf044951b43b3f4f493491e9b381624b
Author: ammirate <cesarano2607@gmail.com>
Date:   Thu Apr 27 14:19:03 2017 +0200

    documentation: enable schema docs and napoleon
    
    * This integrates the jsonschema2rst tool to generate automatically
    documentation from the schemas.
    * Also enables napoleon to have google-style docstrings in python code
    * Closes #73 and #107
09:37 AM ~/Work/repos/per_user/inspirehep/inspire-schemas<venv:inspire-schemas>  (master ↑·3|✔) 
dcaro@pcrcssis001$ ll^C
09:37 AM ~/Work/repos/per_user/inspirehep/inspire-schemas<venv:inspire-schemas>  (master ↑·3|✔) 
dcaro@pcrcssis001$ pip install -e .
Obtaining file:///home/dcaro/Work/repos/per_user/inspirehep/inspire-schemas
Requirement already satisfied: autosemver in /home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages (from inspire-schemas==40.0.2)
Requirement already satisfied: jsonschema in /home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages (from inspire-schemas==40.0.2)
Requirement already satisfied: idutils in /home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages (from inspire-schemas==40.0.2)
Requirement already satisfied: pyyaml in /home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages (from inspire-schemas==40.0.2)
Requirement already satisfied: six in /home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages (from inspire-schemas==40.0.2)
Requirement already satisfied: dulwich in /home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages (from autosemver->inspire-schemas==40.0.2)
Requirement already satisfied: functools32; python_version == "2.7" in /home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages (from jsonschema->inspire-schemas==40.0.2)
Requirement already satisfied: isbnid-fork>=0.4.4 in /home/dcaro/.virtualenvs/inspire-schemas/lib/python2.7/site-packages (from idutils->inspire-schemas==40.0.2)
Installing collected packages: inspire-schemas
  Found existing installation: inspire-schemas 40.0.2
    Uninstalling inspire-schemas-40.0.2:
      Successfully uninstalled inspire-schemas-40.0.2
  Running setup.py develop for inspire-schemas
Successfully installed inspire-schemas
09:37 AM ~/Work/repos/per_user/inspirehep/inspire-schemas<venv:inspire-schemas>  (master ↑·3|✔) 
dcaro@pcrcssis001$ ./run-tests.sh 
Skipped 2 files
======================================================== test session starts =========================================================
platform linux2 -- Python 2.7.13, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /home/dcaro/.virtualenvs/inspire-schemas/bin/python2
cachedir: .cache
rootdir: /home/dcaro/Work/repos/per_user/inspirehep/inspire-schemas, inifile:
plugins: pep8-1.0.6, cov-2.4.0
collected 51 items 

tests/integration/test_builders.py PASSED
tests/integration/test_builders.py::test_literature_builder_valid_record PASSED
tests/integration/test_setup.py PASSED
tests/integration/test_setup.py::test_json_files_are_created_from_yml_files PASSED
tests/integration/test_validate.py PASSED
tests/integration/test_validate.py::test_schemas_validate[experiments] PASSED
tests/integration/test_validate.py::test_schemas_validate[jobs] PASSED
tests/integration/test_validate.py::test_schemas_validate[hep] PASSED
tests/integration/test_validate.py::test_schemas_validate[journals] PASSED
tests/integration/test_validate.py::test_schemas_validate[conferences] PASSED
tests/integration/test_validate.py::test_schemas_validate[institutions] PASSED
tests/integration/test_validate.py::test_schemas_validate[authors] PASSED
tests/integration/test_validate.py::test_schemas_validate_negative[experiments] PASSED
tests/integration/test_validate.py::test_schemas_validate_negative[jobs] PASSED
tests/integration/test_validate.py::test_schemas_validate_negative[hep] PASSED
tests/integration/test_validate.py::test_schemas_validate_negative[journals] PASSED
tests/integration/test_validate.py::test_schemas_validate_negative[conferences] PASSED
tests/integration/test_validate.py::test_schemas_validate_negative[institutions] PASSED
tests/integration/test_validate.py::test_schemas_validate_negative[authors] PASSED
tests/unit/test_api.py PASSED
tests/unit/test_api.py::test_validate_raises_if_no_schema_key PASSED
tests/unit/test_api.py::test_validate_ok_if_schema_key PASSED
tests/unit/test_api.py::test_validate_ok_if_schema_param PASSED
tests/unit/test_hep_builder.py PASSED
tests/unit/test_hep_builder.py::test_is_citeable[True-formdata0] PASSED
tests/unit/test_hep_builder.py::test_is_citeable[True-formdata1] PASSED
tests/unit/test_hep_builder.py::test_is_citeable[False-formdata2] PASSED
tests/unit/test_hep_builder.py::test_is_citeable[False-formdata3] PASSED
tests/unit/test_hep_builder.py::test_append_to PASSED
tests/unit/test_utils.py PASSED
tests/unit/test_utils.py::test_get_schema_path_positive[relative simple path] PASSED
tests/unit/test_utils.py::test_get_schema_path_positive[relative subfolder path] PASSED
tests/unit/test_utils.py::test_get_schema_path_positive[relative simlple path without extension] PASSED
tests/unit/test_utils.py::test_get_schema_path_positive[absolute path] PASSED
tests/unit/test_utils.py::test_get_schema_path_positive[dotted_path] PASSED
tests/unit/test_utils.py::test_get_schema_path_positive[full url] PASSED
tests/unit/test_utils.py::test_get_schema_path_positive[too many dotted relative path] PASSED
tests/unit/test_utils.py::test_get_schema_path_positive[too many dotted full url] PASSED
tests/unit/test_utils.py::test_get_schema_path_negative[non existing path] PASSED
tests/unit/test_utils.py::test_get_schema_path_negative[existing malicious path] PASSED
tests/unit/test_utils.py::test_local_ref_resolver_proxied PASSED
tests/unit/test_utils.py::test_local_ref_resolver_adapted PASSED
tests/unit/test_utils.py::test_load_schema_with_schema_key PASSED
tests/unit/test_utils.py::test_load_schema_without_schema_key PASSED
inspire_schemas/__init__.py PASSED
inspire_schemas/api.py PASSED
inspire_schemas/builders.py PASSED
inspire_schemas/errors.py PASSED
inspire_schemas/utils.py PASSED
inspire_schemas/records/__init__.py PASSED
inspire_schemas/records/elements/__init__.py PASSED

---------- coverage: platform linux2, python 2.7.13-final-0 ----------
Name                                           Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------
inspire_schemas/__init__.py                        1      0   100%
inspire_schemas/api.py                             4      0   100%
inspire_schemas/builders.py                      212     11    95%   53, 124, 128, 187, 410-411, 424, 649-652
inspire_schemas/errors.py                         11      0   100%
inspire_schemas/records/__init__.py                1      1     0%   25
inspire_schemas/records/elements/__init__.py       1      1     0%   25
inspire_schemas/utils.py                          65      7    89%   157-165, 181
----------------------------------------------------------------------------
TOTAL                                            295     20    93%


===================================================== 51 passed in 2.87 seconds ======================================================
experiments.yml                         OK
authors.json                            OK
data.json                               OK
institutions.json                       OK
authors.yml                             OK
conferences.json                        OK
journals.yml                            OK
institutions.yml                        OK
jobs.yml                                OK
conferences.yml                         OK
experiments.json                        OK
hep.yml                                 OK
journals.json                           OK
data.yml                                OK
jobs.json                               OK
hep.json                                OK
inspire_field.yml                       OK
title.json                              OK
reference.json                          OK
source.json                             OK
cnum.yml                                OK
language_code.json                      OK
material.json                           OK
language_code.yml                       OK
country_code.yml                        OK
sourced_value.yml                       OK
acquisition_source.yml                  OK
cnum.json                               OK
related_record.yml                      OK
arxiv_categories.json                   OK
material.yml                            OK
related_record.json                     OK
country_code.json                       OK
source.yml                              OK
reference.yml                           OK
json_reference.json                     OK
url.json                                OK
url.yml                                 OK
rank.yml                                OK
json_reference.yml                      OK
degree_type.json                        OK
rank.json                               OK
document_type.json                      OK
contact.yml                             OK
address.yml                             OK
acquisition_source.json                 OK
address.json                            OK
contact.json                            OK
title.yml                               OK
inspire_field.json                      OK
sourced_value.json                      OK
id.json                                 OK
id.yml                                  OK
document_type.yml                       OK
degree_type.yml                         OK
arxiv_categories.yml                    OK
Index created.
```
As you can see, it processes both, yml and json, generating an api page with duplicated entries:


screenshot from 2017-06-27 09-39-26

**Comment by david-caro** [comment]
found it xd, the issue is an 'old' jsonschemas2rst installed (0.0.113, from a strange branch). You can merge without problems 👍

**Pull request merged**